### PR TITLE
If const cond

### PR DESCRIPTION
**Pull request by scxiao**
special case (constant condition input) for the if operator...

### CODE REVIEW

File: src/onnx/parse_if.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            const auto& else_graph = info.attributes.at("else_branch").g();
+            parser.parse_graph(mdl, else_graph);
+        }
+
+        // inputs of the return instruction are that of the output of the
+        // if instruction
+        instruction_ref ret_ins = std::prev(mdl->end());
+        auto outputs            = ret_ins->inputs();
+        mdl->remove_instruction(ret_ins);
```
Comment (pfultz2) [inline src/onnx/parse_if.cpp:54]: Should we add an assert that this instruction is `@return`?

File: src/onnx/include/migraphx/onnx/onnx_parser.hpp [hunk 2]
```diff
@@ -27,7 +27,7 @@ struct onnx_parser
         attribute_map attributes{};
         std::size_t num_outputs = 1;
         std::string name        = "";
-        module* mm              = nullptr;
+        module* mdl             = nullptr;
```
Comment (pfultz2) [inline src/onnx/include/migraphx/onnx/onnx_parser.hpp:30]: I think it would be better to name it `mod`.

**Comment by scxiao** [comment]
@pfultz2 When multiple modules run in parallel using the onnx_test_runner, looks like there is a race condition happened. The error info is:
```
Test command: /onnxruntime/build/Linux/Release/onnx_test_runner "/onnxruntime/cmake/external/onnx/onnx/backend/test/data/pytorch-operator"
Test timeout computed to be: 3600

2021-02-02 23:55:21.873167475 [E:onnxruntime:Default, testcase_driver.cc:39 RunParallel] Running tests in parallel: at most 24 models at any time3: 2021-02-02 23:55:22.055331458 [E:onnxruntime:Default, testcase_driver.cc:61 RunModelsAsync] Running tests finished. Generating report /src/external/hip-on-vdi/rocclr/hip_code_object.cpp:92: guarantee(false && "hipErrorNoBinaryForGpu: Coudn't find binary for current devices!")
3/3 Test #3: onnx_test_pytorch_operator .......Child aborted***Exception:   0.48 sec
```
Do you have an idea what is the reason for that?

**Comment by scxiao** [comment]
@pfultz2 @kahmed10 @turneram Could you please help review this PR when you  get a chance?

**Review by kahmed10 — APPROVED** [review]

File: src/onnx/parse_if.cpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        // cond is not constant, need to create sub_modules
+        if(cond_arg.empty())
+        {
+            MIGRAPHX_THROW(
+                "PARSE_IF: current implementation requires condition input to be constant!");
+        }
+
+        std::vector<bool> vec_conds;
+        cond_arg.visit([&](auto s) { vec_conds.assign(s.begin(), s.end()); });
+        if(vec_conds.size() != 1)
```
Comment (pfultz2) [inline src/onnx/parse_if.cpp:31]: This should just do `if(cond_arg->get_shape().elements() != 1)`.

File: src/onnx/parse_if.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        migraphx::argument cond_arg = args.front()->eval();
+        // cond is not constant, need to create sub_modules
+        if(cond_arg.empty())
+        {
+            MIGRAPHX_THROW(
+                "PARSE_IF: current implementation requires condition input to be constant!");
+        }
+
+        std::vector<bool> vec_conds;
+        cond_arg.visit([&](auto s) { vec_conds.assign(s.begin(), s.end()); });
```
Comment (pfultz2) [inline src/onnx/parse_if.cpp:30]: You should just do `cond_arg.at<bool>()` to get a boolean.

File: src/onnx/parse_if.cpp (hunk 3 above)
Comment (scxiao) [reply src/onnx/parse_if.cpp]: fixed.

File: src/onnx/parse_if.cpp (hunk 4 above)
Comment (scxiao) [reply src/onnx/parse_if.cpp]: fixed

**Review by turneram — APPROVED** [review]

**Review by pfultz2 — APPROVED** [review]

**Comment by scxiao** [comment]
I think this PR can merge now. The code coverage check failed, but the missing lines are all for throwing exceptions.

**Pull request merged**